frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,381 @@ pub mod pallet {
+		/// - `ending_block`: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// - `locked`: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get(), T::MaxVestingSchedules::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let schedules = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 =
+				*schedules.get(schedule1_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let schedule2 =
+				*schedules.get(schedule2_index).ok_or(Error::<T>::ScheduleIndexOutOfBounds)?;
+			let merge_action = VestingAction::Merge(schedule1_index, schedule2_index);
+
+			// The length of `schedules` decreases by 2 here since we filter out 2 schedules.
+			// Thus we know below that we can insert the new merged schedule without error (assuming
+			// initial state was valid).
+			let (mut schedules, mut locked_now) =
+				Self::report_schedule_updates(schedules.to_vec(), merge_action);
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			let now = <frame_system::Pallet<T>>::block_number();
+			if let Some(new_schedule) = Self::merge_vesting_info(now, schedule1, schedule2)? {
+				// Merging created a new schedule so we:
+				// 1) need to add it to the accounts vesting schedule collection,
+				schedules.push(new_schedule);
+				// (we use `locked_at` in case this is a schedule that started in the past)
+				let new_schedule_locked = new_schedule.locked_at::<T::BlockNumberToBalance>(now);
+				// 2) update the locked amount to reflect the schedule we just added,
+				locked_now = locked_now.saturating_add(new_schedule_locked);
+				// and 3) deposit an event.
+				Self::deposit_event(Event::<T>::MergedScheduleAdded(
+					new_schedule.locked(),
+					new_schedule.per_block(),
+					new_schedule.starting_block(),
+				));
+			} // In the None case there was no new schedule to account for.
+
+			debug_assert!(
+				locked_now > Zero::zero() && schedules.len() > 0 ||
+					locked_now == Zero::zero() && schedules.len() == 0
+			);
```

Comment:
If we move the merge logic into the function, then we can also move this into the function and sanity check any returned values.

Reply:
4d410f7